org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/DecompilerManager.java

```diff
@@ -0,0 +1,192 @@
+						String message = "Invalid extension to " + EXTENSION_POINT_ID + ". Must implement org.eclipse.jdt.ls.core.internal.IDecompiler";
+						JavaLanguageServerPlugin.logError(message);
+						return null;
+					}
+				} catch (CoreException e) {
+					JavaLanguageServerPlugin.logException("Unable to create decompiler ", e);
+					return null;
+				}
+			}
+			return decompiler;
+		}
+	}
+
+	private static Set<DecompilerDescriptor> descriptors;
+
+	private static synchronized Set<DecompilerDescriptor> getDecompilerDescriptors() {
+		if (descriptors == null) {
+			IConfigurationElement[] elements = Platform.getExtensionRegistry().getConfigurationElementsFor(EXTENSION_POINT_ID);
+			descriptors = Stream.of(elements).map(e -> new DecompilerDescriptor(e)).collect(Collectors.toSet());
+		}
+		return descriptors;
+	}
+
+	/**
+	 * Decompile the given class file if possible
+	 *
+	 * @param classFile
+	 *            the class file to decompile
+	 * @param monitor
+	 * @return the decompiled code if successful, or <code>null</code> if
+	 *         unsuccessful, or an empty string if canceled
+	 */
+	public String decompile(IClassFile classFile, IProgressMonitor monitor) {
+		IDecompiler decompiler = getDecompiler();
+		if (decompiler == null) {
+			JavaLanguageServerPlugin.logError("Unable to load decompiler class");
+			return null;
+		}
+
+		if (monitor.isCanceled()) {
+			return "";
+		}
+
+		final Map<String, Object> configuration = getConfiguration();
+		final String[] resultValues = new String[1];
+		SafeRunner.run(new ISafeRunnable() {
+			@Override
+			public void run() throws Exception {
+				resultValues[0] = decompiler.decompile(classFile, configuration, monitor);
+			}
+
+			@Override
+			public void handleException(Throwable ex) {
+				IStatus status = new Status(IStatus.ERROR, JavaLanguageServerPlugin.PLUGIN_ID, IStatus.OK, "Error calling decompiler", ex);
+				JavaLanguageServerPlugin.log(status);
+			}
+		});
+		return prependHeader(resultValues[0]);
+	}
+
+	@SuppressWarnings("unchecked")
+	private Map<String, Object> getConfiguration() {
+		String decompilerId = preferenceManager.getPreferences().getDecompilerId();
+		Map<String, Object> configuration = preferenceManager.getPreferences().getDecompilerConfiguration();
+		if (configuration != null && configuration.get(decompilerId) instanceof Map) {
+			configuration = (Map<String, Object>) configuration.get(decompilerId);
+		}
+		return configuration;
+	}
+
+	private IDecompiler getDecompiler() {
+		if (this.decompiler != null) {
+			return this.decompiler;
+		}
+
+		Set<DecompilerDescriptor> handlers = getDecompilerDescriptors();
+		if (handlers.isEmpty()) {
+			JavaLanguageServerPlugin.logError("No decompilers found");
+			return null;
+		}
+
+		String decompilerId = preferenceManager.getPreferences().getDecompilerId();
+		Optional<DecompilerDescriptor> handler = handlers.stream().filter(d -> decompilerId == null || decompilerId.equals(d.id)).findFirst();
+		if (!handler.isPresent()) {

```

Comment:
if there's no matching decompiler, we should log it and fall back on the disassembler